pkg/geo/geos/geos_unix.go

```diff
@@ -44,29 +44,77 @@ var geosOnce struct {
 
 // EnsureInit attempts to start GEOS if it has not been opened already
 // and returns an error if the CR_GEOS  is not valid.
-func EnsureInit(errDisplay EnsureInitErrorDisplay) error {
-	_, err := ensureInit(errDisplay)
+func EnsureInit(errDisplay EnsureInitErrorDisplay, flagGEOSLocationValue string) error {
+	_, err := ensureInit(errDisplay, flagGEOSLocationValue)
 	return err
 }
 
-// ensureInit behaves as described in EnsureInit, but also returns the GEOS
+// ensureInitInternal ensures initialization has been done, always displaying
+// errors privately and not assuming a flag has been set if initialized
+// for the first time.
+func ensureInitInternal() (*C.CR_GEOS, error) {
+	return ensureInit(EnsureInitErrorDisplayPrivate, "")
+}
+
+// ensureInits behaves as described in EnsureInit, but also returns the GEOS
 // C object which should be hidden from the public eye.
-func ensureInit(errDisplay EnsureInitErrorDisplay) (*C.CR_GEOS, error) {
+func ensureInit(
+	errDisplay EnsureInitErrorDisplay, flagGEOSLocationValue string,
+) (*C.CR_GEOS, error) {
 	geosOnce.once.Do(func() {
-		geosOnce.geos, geosOnce.err = initGEOS(defaultGEOSLocations)
+		geosOnce.geos, geosOnce.err = initGEOS(findGEOSLocations(flagGEOSLocationValue))
 	})
 	if geosOnce.err != nil && errDisplay == EnsureInitErrorDisplayPublic {
 		return nil, errors.Newf("geos: this operation is not available")
 	}
 	return geosOnce.geos, geosOnce.err
 }
 
-// defaultGEOSLocations contains a list of locations where GEOS is expected to exist.
-// TODO(otan): make this configurable by flags.
-// TODO(otan): put mac / linux locations
-var defaultGEOSLocations []string
+// findGEOSLocations returns the default locations where GEOS is installed.
+func findGEOSLocations(flagGEOSLocationValue string) []string {
+	var ext string
+	switch runtime.GOOS {
+	case "darwin":
+		ext = "dylib"
+	default:
+		ext = "so"
+	}
+	if flagGEOSLocationValue != "" {
+		// Try both the location itself, as well as adding the libgeos_c extension
+		// at the end in case our user assumes it's just a directory.
+		return []string{
+			flagGEOSLocationValue,
+			filepath.Join(flagGEOSLocationValue, "libgeos_c."+ext),
+		}
+	}
+	var locDirs []string
+	switch runtime.GOOS {
+	case "darwin":
+		locDirs = []string{
+			"/usr/local/lib",
+			"/usr/lib",
+		}
+	default:
+		locDirs = []string{
+			"/usr/local/lib",
+			"/usr/lib/x86_64-linux-gnu",
+			"/usr/lib/i386-linux-gnu",
+			"/usr/lib",
+		}
+	}
+	var locs []string
+	locs = append(locs, findGEOSLocationsInParentingDirectories(ext)...)
+	for _, loc := range locDirs {
+		locs = append(locs, filepath.Join(loc, "libgeos_c."+ext))
```

Comment:
Please log the directory that was selected when auto-detection succeeds. This is necessary for troubleshooting.

Reply:
i'll keep them, separate for now!